### DISCUSSION TITLE
Remote hardware attestation

### DISCUSSION THREAD

**Discussion by TommyTran732** [Divested-Mobile/DivestOS-Build#46] [Ideas]
GrapheneOS has their Auditor app which can do both local attestation using another Android 8+ device and remote attestation with attestation.app. Having continuous device monitoring is actually quite nice. It would be awesome if DivestOS can have something like this.

If you are okay with the idea, we could:
1. Ask GrapheneOS to include DivestOS devices in their client and server. They have a few requirements like the OS must at least uphold the Android security model, have verified boot, not have signature spoofing, etc. None of which seems to be a problem with DivestOS as far as I know.
2. Fork the auditor app and setup DivestOS's own attestation server. If you do this, then you have full control over the infrastructure and don't have to rely on another project for support. It will, however, get annoying if other Android derivatives start doing the same thing, as there will be a bunch of auditor forks and doing local attestation between 2 different android derivatives would require installing 2 different versions of the auditor.

**Comment by flawedworld** [comment]
@SkewedZeppelin Please get in touch if you are interested in this, would be happy to help!

**Reply by SkewedZeppelin** [reply]
I'm not against this feature.
And it would be very nice to see.
The issues I have are:
- I see no easy way to actually generate the data and there is very little documentation on it.
  - there is just the option to submit sample data
  - from my understanding you can't even trivially locally compile it, because it only works when it is signed with the official key
- DivestOS has few tested working devices that meet the criteria
  - for example I only have a 1st gen Pixel, 2 and 3 have been tested by others, 3a is very outdated because I can't compile 11 for it, 4/5 don't boot, and 6 isn't supported yet
  - OnePlus 6 and 6T for example currently use the official stock vendor partition

**Reply by flawedworld** [reply]
Device support would be a matter of doing the following for each device. Example values have been given for a Pixel 6 running the stock OS.

- obtaining verified boot key SHA256's of DivestOS ("0F6E75C80183B5DEC074B0054D4271E99389EBE4B136B0819DE1F150BA0FF9D7")
- obtaining device model name ("Google Pixel 6")
- obtaining attestation version ("4")
- obtaining keymaster version ("41")
- if they use the older rollback resistance API (Pixel 2 and older) ("false")
- if they have modern per-user encryption (basically near-everything)  ("true")
- if they have strongbox support ("true")

You can get attestation/keymaster versions quickly by compiling a dev build of Auditor, adding junk info (make them say 184 or whatever) and getting the real values via the logs it will produce. The TEE and strongbox based attestation may produce different values and you can force TEE based attestation by changing: https://github.com/GrapheneOS/Auditor/blob/f90109d2e2ddcee03ca3095f19ace7a826810b38/app/src/main/java/app/attestation/auditor/AttestationProtocol.java#L121

Here is the work we did for adding the Pixel 6 and 6 Pro for example:
https://github.com/GrapheneOS/Auditor/commit/c5f9827b3868537c041b790174376161ea2bba97
https://github.com/GrapheneOS/Auditor/commit/4e377bcd0e4d138a8cf5babd0823b4c12bee4157

https://github.com/GrapheneOS/AttestationServer/commit/17ee0ff82cb56033d0ba0da0a33dd8a433aa903f
https://github.com/GrapheneOS/AttestationServer/commit/27d6a13f9a03f553b2bbb792e1ff0c1e86cceed1

If you were to go about implementing support in DivestOS, the easiest route would be adding support for devices which you have access to, DivestOS has production OS support for and we already have infrastructure for in Auditor/AttestationServer in my opinion. You can test Auditor easily locally with 2 Android devices with your own builds and so long as you update the key within the code. If you nail that then AttestationServer becomes easier to do.

**Reply by SkewedZeppelin** [reply]
OK
So I first created a list of the verified boot keys.
And here they are for reference: https://divestos.org/misc/verified_boot_public_key_hashes.txt
I verified they match the boot screen on mata and fajita. A user in my chat room also verified guacamole.
I could not figure out how to derive the one on my sailfish. Looks like SHA-1 but didn't match my script.

I then added fajita to Auditor, but immediately slammed into an error and can also reproduce it with the official Auditor app.
This appears to be an [common](https://forum.xda-developers.com/t/solved-android-attest-key-lost.3829945/) issue where the attestation key can be lost?
```
02-27 22:29:41.024   577   577 E KeyMasterHalDevice: Attest key send cmd failed
02-27 22:29:41.024   577   577 E KeyMasterHalDevice: ret: 0
02-27 22:29:41.024   577   577 E KeyMasterHalDevice: resp->status: -10003
02-27 22:29:41.024  1122  5208 E keystore: Keymaster reported error: -10003
02-27 22:29:41.024  1122  5208 E keystore: NOTE: This is an error in the vendor specific error range.
02-27 22:29:41.024  1122  5208 E keystore:       Refer to the vendor of the implementation for details.
02-27 22:29:41.024  1122  5208 E keystore:       Implementation name: Keymaster HAL: 4
02-27 22:29:41.024  1122  5208 E keystore:       Vendor name:         QTI
02-27 22:29:41.024  1122  5208 E keystore:       MajorVersion:        
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: attestation generation error
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: java.io.IOException: java.security.ProviderException: Failed to generate attestation certificate chain
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at app.attestation.auditor.AttestationProtocol.generateKeyPair(AttestationProtocol.java:1438)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at app.attestation.auditor.AttestationProtocol.generateSerialized(AttestationProtocol.java:1270)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at app.attestation.auditor.GenerateAttestationService.onHandleIntent(GenerateAttestationService.java:62)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:77)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at android.os.Handler.dispatchMessage(Handler.java:106)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at android.os.Looper.loop(Looper.java:223)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at android.os.HandlerThread.run(HandlerThread.java:67)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: Caused by: java.security.ProviderException: Failed to generate attestation certificate chain
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at android.security.keystore.AndroidKeyStoreKeyPairGeneratorSpi.getAttestationChain(AndroidKeyStoreKeyPairGeneratorSpi.java:611)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at android.security.keystore.AndroidKeyStoreKeyPairGeneratorSpi.createCertificateChain(AndroidKeyStoreKeyPairGeneratorSpi.java:498)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at android.security.keystore.AndroidKeyStoreKeyPairGeneratorSpi.generateKeyPair(AndroidKeyStoreKeyPairGeneratorSpi.java:475)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at java.security.KeyPairGenerator$Delegate.generateKeyPair(KeyPairGenerator.java:727)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at app.attestation.auditor.AttestationProtocol.generateKeyPair(AttestationProtocol.java:1436)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	... 6 more
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: Caused by: android.security.KeyStoreException: -10003
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	at android.security.KeyStore.getKeyStoreException(KeyStore.java:1301)
02-27 22:29:41.027  1907  5254 E GenerateAttestationService: 	... 11 more
```

I can't really continue, as I don't have any other supported devices.

**Reply by flawedworld** [reply]
Seems like they screwed it up. Typical OnePlus.

**Reply by flawedworld** [reply]
A lot of OEM's in particular incompetent ones like OnePlus ending up breaking attestation APIs with updates (or it never worked) and I highly suspect this is the case here.

**Comment by flawedworld** [comment]
Due to a bug in some OEM implementations of the Android keystore, you will need to copy this for some devices  https://github.com/GrapheneOS/platform_system_security/commit/74760ab693d7cd4e64d193f5d901173ab5b34f82

This affects the Pixel 3 and 3 XL at least

**Reply by SkewedZeppelin** [reply]
I've picked that: https://github.com/Divested-Mobile/DivestOS-Build/commit/2de5521a7c851381e9b716fc2ff46ef14e316258

Thanks!

**Reply by flawedworld** [reply]
This is technically disabling a defense in depth feature, so it only should be applied to affected devices

**Reply by SkewedZeppelin** [reply]
Yes, I've noted it.
The issue is I can't know exactly which are actually impacted.

**Reply by flawedworld** [reply]
Rough guess would be anything with firmware from ~November 2021-March 2022 or newer should be unaffected

**Reply by flawedworld** [reply]
https://android-review.googlesource.com/c/platform/system/security/+/1930568 is the correct way to deal with this

**Reply by SkewedZeppelin** [reply]
Yes, I had already dropped the revert patch a day later after LineageOS pulled it in:
https://review.lineageos.org/c/LineageOS/android_system_security/+/327113
